File: mobile/ios/PairMobileKit/Sources/PairMobileKit/PairAPIError.swift
import Foundation

/// Normalized error payload returned by Pair API endpoints.
public struct PairAPIErrorPayload: Codable, Equatable, Sendable {
	public let code: String?
	public let message: String?
	public let details: [String: [String]]?

	/// Initializes the backend error payload.
	public init(code: String? = nil, message: String? = nil, details: [String: [String]]? = nil) {
		self.code = code
		self.message = message
		self.details = details
	}
}

/// Common client errors for Pair-based native apps.
public enum PairAPIError: Error, Equatable, Sendable {
	case invalidBaseURL
	case invalidResponse
	case server(statusCode: Int, payload: PairAPIErrorPayload?)
	case decoding
}

public extension PairAPIError {

	/// Indicates whether the backend rejected the current Bearer session.
	var isAuthenticationFailure: Bool {
		if case .server(let statusCode, _) = self {
			return statusCode == 401
		}

		return false
	}
}

struct PairAPIErrorEnvelope: Decodable {
	let error: PairAPIErrorPayload?

	private enum CodingKeys: String, CodingKey {
		case error
		case code
		case message
		case details
	}

	/// Decodes both the nested mobile envelope and Pair's established flat API error payload.
	init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)

		if let nested = try? container.decode(PairAPIErrorPayload.self, forKey: .error) {
			error = nested
			return
		}

		let code = try container.decodeIfPresent(String.self, forKey: .code)
		let message = try container.decodeIfPresent(String.self, forKey: .message)
			?? container.decodeIfPresent(String.self, forKey: .error)
		let details = try container.decodeIfPresent([String: [String]].self, forKey: .details)

		error = nil == code && nil == message && nil == details
			? nil
			: PairAPIErrorPayload(code: code, message: message, details: details)
	}
}

File: mobile/android/PairMobileAndroid/src/main/kotlin/dev/pair/mobile/android/PairApiError.kt
package dev.pair.mobile.android

import kotlinx.serialization.Serializable

/** Normalized error payload returned by Pair API endpoints. */
@Serializable
data class PairApiErrorPayload(
    val code: String? = null,
    val message: String? = null,
    val details: Map<String, List<String>>? = null
)

/** Common client errors for Pair-based Android apps. */
sealed class PairApiException(message: String? = null, cause: Throwable? = null) : Exception(message, cause) {

    /** The configured API base URL or a request URL is invalid. */
    data object InvalidUrl : PairApiException("Invalid Pair API URL.")

    /** The transport returned a response that cannot be consumed safely. */
    data object InvalidResponse : PairApiException("Invalid Pair API response.")

    /** The backend returned a non-successful HTTP status code. */
    data class Server(
        val statusCode: Int,
        val payload: PairApiErrorPayload?
    ) : PairApiException(payload?.message ?: "Pair API request failed with HTTP $statusCode.")

    /** The response body does not match the expected JSON contract. */
    class Decoding(cause: Throwable) : PairApiException("Pair API response decoding failed.", cause)

    /** The underlying HTTP transport failed before a valid HTTP response was available. */
    class Transport(cause: Throwable) : PairApiException("Pair API transport failed.", cause)

    /** Indicates whether the backend rejected the current Bearer session. */
    val isAuthenticationFailure: Boolean
        get() = this is Server && statusCode == 401
}

@Serializable
internal data class PairApiErrorEnvelope(
    val error: PairApiErrorPayload? = null
)

